code/game/gamemodes/changeling/evolution_menu.dm

```diff
@@ -1,300 +1,88 @@
-				</td>
-			</tr>
-	</table>
-
-	"}
-
-	//player table header
-	dat += {"
-		<span id='maintable_data_archive'>
-		<table width='560' align='center' cellspacing='0' cellpadding='5' id='maintable_data'>"}
-
-	var/i = 1
-	for(var/datum/action/changeling/cling_power in GLOB.sting_paths)
-
-		if(cling_power.dna_cost <= 0) //Let's skip the crap we start with. Keeps the evolution menu uncluttered.
+	/// Which UI view will be displayed. Compact mode will show only ability names, and will leave out their descriptions and helptext.
+	var/view_mode = EXPANDED_MODE
+	/// A list containing the names of bought changeling abilities. For use with the UI.
+	var/list/purchased_abilities = list()
+	/// A list containing every purchasable changeling ability. Includes its name, description, helptext and cost.
+	var/static/list/ability_list = list()
+
+/datum/action/changeling/evolution_menu/Grant(mob/M)
+	. = ..()
+	if(length(ability_list))
+		return // List is already populated.
+
+	for(var/action in subtypesof(/datum/action/changeling))
+		var/datum/action/changeling/C = action
+		if(initial(C.dna_cost) <= 0) // Filter out innate abilities like DNA sting, Evolution menu, etc.
 			continue
+		ability_list += list(list(
+			"name" = initial(C.name),
+			"description" = initial(C.desc),
+			"helptext" = initial(C.helptext),
+			"cost" = initial(C.dna_cost)
+		))
 
-		var/ownsthis = changeling.has_sting(cling_power)
-
-		var/color
-		if(ownsthis)
-			if(i%2 == 0)
-				color = "#d8ebd8"
-			else
-				color = "#c3dec3"
-		else
-			if(i%2 == 0)
-				color = "#f2f2f2"
-			else
-				color = "#e6e6e6"
-
-
-		dat += {"
-
-			<tr id='data[i]' name='[i]' onClick="addToLocked('item[i]','data[i]','notice_span[i]')">
-				<td align='center' bgcolor='[color]'>
-					<span id='notice_span[i]'></span>
-					<a id='link[i]'
-					onmouseover='expand("item[i]","[cling_power.name]","[cling_power.desc]","[cling_power.helptext]","[cling_power]",[ownsthis])'
-					>
-					<b id='search[i]'>Evolve [cling_power][ownsthis ? " - Purchased" : (cling_power.req_dna>changeling.absorbedcount ? " - Requires [cling_power.req_dna] absorptions" : " - Cost: [cling_power.dna_cost]")]</b>
-					</a>
-					<br><span id='item[i]'></span>
-				</td>
-			</tr>
-
-		"}
-
-		i++
-
-
-	//player table ending
-	dat += {"
-		</table>
-		</span>
-
-		<script type='text/javascript'>
-			var maintable = document.getElementById("maintable_data_archive");
-			var complete_list = maintable.innerHTML;
-		</script>
-	</body></html>
-	"}
-	return dat
-
-
-/datum/action/changeling/evolution_menu/Topic(href, href_list)
-	..()
-	if(!(iscarbon(usr) && usr.mind && usr.mind.changeling))
+/datum/action/changeling/evolution_menu/Trigger()
+	tgui_interact(owner)
+
+/datum/action/changeling/evolution_menu/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_always_state)
```

Comment:
Unsure about the `always` state. Should changelings be able to modify their abilities even when dead?

Reply:
To be honest, I used something that worked for the time being, because some of the other states I tried weren't allowing me to open the UI, so I chose this. If there's a better state I can use here, I'll switch to that one. 

You could also say that changelings aren't really dead I guess. Even when they're "dead", they can revive.